doc: flpr and building for nrf54l15 (#17443)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/nrf54l/vpr_flpr.rst

.. _vpr_flpr_nrf54l:

Working with the FLPR core
##########################

> The nRF54L15 SoC has a dedicated RISC-V CPU (VPR), a fast lightweight peripheral processor (FLPR) designed for software-defined peripherals (SDP).

Reviewer: ```suggestion
The nRF54L15 SoC has a dedicated RISC-V CPU (VPR), a fast lightweight peripheral processor (FLPR) optimized for software-defined peripherals (SDP).
```

> This core currently supports the following peripherals:

Reviewer 2: "core" is ambiguous. Unfortunately, IPS uses this term to mean something else.
```suggestion
This CPU currently supports the following peripherals:
```

> This core currently supports the following peripherals:

Reviewer 2: Which CPU supports these peripherals? Are SDPs emulating these peripherals available? Coes Application CPU support these? Of are device drivers for these available in FLPR? Or something else :)

Reviewer: How about
`Following peripherals are available to be utilized on the FLPR core, using appropriate Zephyr Device Driver API:`

* GPIO
* GPIOTE
* GRTC
* TWIM
* UARTE
* VPR

> Running the FLPR core

Reviewer 2: ```suggestion
Running FLPR
```

> Running the FLPR core
> *********************
> 
> VPR is optimized for implementing SDP.
> The software-defined peripheral running on VPR does not start on its own, but must be started by the application core processor.
> 
> .. To be outlined how

Reviewer: ```suggestion
Running the FLPR core
*********************

General usage
-------------
FLPR can be used as a generic core, running full Zephyr kernel. In this case, FLPR target can be built similarly to APP core. Note that APP core build needs to include an overlay enabling FLPR coprocessor, so the code that starts FLPR will be included.

Software Defined Peripherals
----------------------------
```
VPR is optimized for implementing Software Defined Peripherals (SDP). Full documentation on SDP can be found here: <TODOLINK>
```

Author: I added this but would this be enough for the user to know how to proceed? Shouldn't we provide some more details on it?

Reviewer: I was thinking about the SDP documentation showing how to proceed in that topic.
When it comes to General usage, this is just using full Zephyr on FLPR, so not sure what we could add here.

Memory allocation
*****************

> If a SDP is active on a VPR CPU, it can lead to increased latency when accessing ``RAM_01``.

Reviewer: ```suggestion
If a VPR CPU is running code, it can lead to increased latency when accessing ``RAM_01``.
```

> Because of this, when SDP is enabled in a project, you should use ``RAM_01`` to store only the VPR code, data, and other non-time-sensitive information.

Reviewer: ```suggestion
Because of this, when FLPR is used in a project, you should use ``RAM_01`` to store only the VPR code, data, and other non-time-sensitive information.
```

> Because of this, when SDP is enabled in a project, you should use ``RAM_01`` to store only the VPR code, data, and other non-time-sensitive information.

Reviewer 2: It's unclear shat "data" do you mean.

```suggestion
Because of this, when SDP is enabled in a project, you should use ``RAM_01`` to store only the VPR code, VPR data, and the Application CPU's non-time-sensitive information.
```

> Conversely, you should use ``RAM_00`` to store data with strict access time requirements such as DMA buffers, and data used in low-latency ISRs.

Reviewer: Should we mention that this is from APP core's perspective? @hubertmis

Reviewer 2: Yes. I would rephrase:
```suggestion
Conversely, you should use ``RAM_00`` to store data with strict access time requirements such as DMA buffers, and Application CPU data used in low-latency ISRs.
```

.. _vpr_flpr_nrf54l:

Working with the FLPR core
##########################

> The nRF54L15 SoC has a dedicated RISC-V CPU (VPR), a fast lightweight peripheral processor (FLPR) optimized for software-defined peripherals (SDP).

Reviewer 2: This sentence can be interpreted as there were two CPUs: VPR for general usage and FLPR for SDPs.

```suggestion
The nRF54L15 SoC has a dedicated VPR CPU (RISC-V architecture), named "fast lightweight peripheral processor" (FLPR) optimized for software-defined peripherals (SDP).
```

> The nRF54L15 SoC has a dedicated VPR CPU (RISC-V architecture), named *fast lightweight peripheral processor* (FLPR) optimized for software-defined peripherals (SDP).

Reviewer: ```suggestion
The nRF54L15 SoC has a dedicated VPR CPU (RISC-V architecture), named *fast lightweight peripheral processor* (FLPR).
```

The following peripherals are available for use with the FLPR core, and can be accessed through the appropriate Zephyr Device Driver API:

* GPIO
* GPIOTE
* GRTC
* TWIM
* UARTE
* VPR

> Running FLPR

Reviewer 3: I would suggest "Using FLPR with Zephyr multithreaded mode".

************

FLPR can function as a generic core, operating under the full Zephyr kernel.
In this configuration, building the FLPR target is similar the application core.
However, the application core build must incorporate an overlay that enables the FLPR coprocessor.
This ensures that the necessary code to initiate FLPR is integrated.

> Software Defined Peripherals

Reviewer 3: I would suggest "Using FLPR with Zephyr threadless mode".

.. _vpr_flpr_nrf54l:

Working with the FLPR core
##########################

.. contents::
   :local:
   :depth: 2

The nRF54L15 SoC has a dedicated VPR CPU (RISC-V architecture), named *fast lightweight peripheral processor* (FLPR).
The following peripherals are available for use with the FLPR core, and can be accessed through the appropriate Zephyr Device Driver API:

* GPIO
* GPIOTE
* GRTC
* TWIM
* UARTE

> * VPR

Reviewer 4: Can you add a short section here to explain how VPR is used to start FLPR? I mean the launcher thing, but I refuse to call it like this. :)

Author: added

.. _vpr_flpr_nrf54l:

Working with the FLPR core
##########################

.. contents::
   :local:
   :depth: 2

The nRF54L15 SoC has a dedicated VPR CPU (RISC-V architecture), named *fast lightweight peripheral processor* (FLPR).
The following peripherals are available for use with the FLPR core, and can be accessed through the appropriate Zephyr Device Driver API:

* GPIO
* GPIOTE
* GRTC
* TWIM
* UARTE
* VPR

Using FLPR with Zephyr multithreaded mode
*****************************************

FLPR can function as a generic core, operating under the full Zephyr kernel.
In this configuration, building the FLPR target is similar the application core.
However, the application core build must incorporate an overlay that enables the FLPR coprocessor.
This ensures that the necessary code to initiate FLPR is integrated.

Using FLPR with Zephyr threadless mode
**************************************

FLPR is optimized for implementing software-defined peripherals (SDP).

> For more information, see <link tba> documentation page.

Reviewer: No, this subchapter could be removed for now.

> For more information, see <link tba> documentation page.

Author: removed

.. _vpr_flpr_nrf54l:

Working with the FLPR core
##########################

.. contents::
   :local:
   :depth: 2

The nRF54L15 SoC has a dedicated VPR CPU (RISC-V architecture), named *fast lightweight peripheral processor* (FLPR).
The following peripherals are available for use with the FLPR core, and can be accessed through the appropriate Zephyr Device Driver API:

* GPIO
* GPIOTE
* GRTC
* TWIM
* UARTE
* VPR

.. _vpr_flpr_nrf54l15_initiating:

Using FLPR with Zephyr multithreaded mode
*****************************************

FLPR can function as a generic core, operating under the full Zephyr kernel.
In this configuration, building the FLPR target is similar to the application core.
However, the application core build must incorporate an overlay that enables the FLPR core.

Bootstrapping the FLPR core
===========================

> The |NCS| provides an FLPR snippet that adds an overlay required for bootstrapping the FLPR core.

Reviewer 4: ```suggestion
The |NCS| provides a FLPR snippet that adds an overlay required for bootstrapping the FLPR core.
```
